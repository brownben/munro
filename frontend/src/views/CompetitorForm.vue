<!--
  Competitor Form

  The form for Creating/ Updating Competitor
-->

<template>
  <div class="view">
    <vue-headful
      v-if="create"
      title="Munro - Create Competitor"
      description
      :head="{
        'meta': {name: 'robots', content:'noindex'},
      }"
    />
    <vue-headful
      v-else
      title="Munro - Edit Competitor"
      description
      :head="{
        'meta': {name: 'robots', content:'noindex'},
      }"
    />
    <div v-if="!notFound">
      <h1 v-if="create" class="text-main text-3xl font-normal font-heading mb-2">Create Competitor</h1>
      <h1 v-if="!create" class="text-main text-3xl font-normal font-heading mb-2">Edit Competitor</h1>
      <!-- @submit on submit via enter key in the last field, .prevent prevents page reload -->
      <form @submit.prevent="submit">
        <text-input v-model.trim="name" label="Name:" />
        <text-input v-model.trim="club" label="Club:" />
        <text-input v-model.trim="ageClass" label="Age Class:" />
        <dropdown-input
          v-model="league"
          :list="leagues.map(league => league.name)"
          label="League:"
        />
        <dropdown-input v-model="course" :list="coursesInLeague" label="Course:" />
        <button v-if="create" class="button-lg">Create Competitor</button>
        <button v-if="!create" class="button-lg">Update Competitor</button>
      </form>
    </div>
    <not-found v-if="notFound" />
  </div>
</template>

<script>
import axios from 'axios'
import DropdownInput from '@/components/DropdownInput'
import TextInput from '@/components/TextInput'
const NotFound = () => import('@/views/NotFound')

export default {
  components: {
    'NotFound': NotFound,
    'DropdownInput': DropdownInput,
    'TextInput': TextInput,
  },

  data: function () {
    return {
      id: '',
      notFound: false,
      create: true,
      leagues: [],
      name: '',
      club: '',
      ageClass: '',
      league: '',
      course: '',
    }
  },

  computed: {
    coursesInLeague: function () {
      const selectedLeague = this.leagues.filter(league => league.name === this.league)
      if (selectedLeague.length > 0) return selectedLeague[0].courses
      else return []
    },
  },

  // On Load
  mounted: function () {
    this.getLeagues()
    if (this.$route.path.includes('edit')) {
      this.create = false
      this.getCompetitorDetails()
    }
  },

  methods: {
    submit: function () {
      if (this.create) this.createCompetitor()
      else this.updateCompetitor()
    },

    getCompetitorDetails: function () {
      return axios.get('/api/competitors/' + this.$route.params.id)
        .then(response => {
          if (!response.data) this.notFound = true
          else {
            this.id = this.$route.params.id
            this.name = response.data.name
            this.club = response.data.club
            this.ageClass = response.data.ageClass
            this.league = response.data.league
            this.course = response.data.course
          }
        })
        .catch(() => this.$messages.addMessage('Problem Getting Competitor Details'))
    },

    getLeagues: function () {
      return axios.get('/api/leagues')
        .then(response => { this.leagues = response.data })
        .catch(() => this.$messages.addMessage('Problem Fetching List of Leagues'))
    },

    validateForm: function () {
      return (this.name !== '' && this.league !== '' && this.course !== '')
    },

    createCompetitor: function () {
      if (this.validateForm()) {
        return axios.post('/api/competitors', {
          name: this.name,
          club: this.club,
          ageClass: this.ageClass,
          league: this.league,
          course: this.course,
        })
          .then(response => this.returnToCompetitorsPage(response))
          .catch(() => this.$messages.addMessage('Error: Problem Creating Competitor - Please Try Again'))
      }
      else this.$messages.addMessage('Please Ensure Name and League Fields are not Blank')
    },

    updateCompetitor: function () {
      if (this.validateForm()) {
        return axios.put('/api/competitors/' + this.id, {
          id: this.id,
          name: this.name,
          club: this.club,
          ageClass: this.ageClass,
          league: this.league,
          course: this.course,
        })
          .then(response => this.returnToCompetitorsPage(response))
          .catch(error => this.$messages.addMessage(error.response.data.message))
      }
      else this.$messages.addMessage('Please Ensure Name and League Fields are not Blank')
    },

    returnToCompetitorsPage: function (response) {
      // Go to league page after successful update/ creation
      this.$messages.addMessage(response.data.message)
      this.$router.push('/competitors')
    },
  },
}

</script>
