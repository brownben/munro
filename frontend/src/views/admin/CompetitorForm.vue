<!--
  Competitor Form

  The form for Creating/ Updating Competitor
-->

<template>
  <div>
    <Meta
      :head="{
        meta: { name: 'robots', content: 'noindex' },
      }"
      :title="`Munro - ${title}`"
      description
    />

    <Layout v-if="!notFound" :title="title">
      <!-- @submit on submit via enter key in the last field, .prevent prevents page reload -->
      <form class="col-span-2" @submit.prevent="submit">
        <TextInput v-model.trim="name" label="Name:" />
        <TextInput v-model.trim="club" label="Club:" class="mt-4" />
        <TextInput v-model.trim="ageClass" label="Age Class:" class="mt-4" />
        <DropdownInput
          v-model="league"
          :list="leagues.map((league) => league.name)"
          label="League:"
          class="mt-4"
        />
        <DropdownInput
          v-model="course"
          :list="coursesInLeague"
          label="Course:"
          class="mt-4"
        />

        <button v-if="create" class="mt-8 button-lg">Create Competitor</button>
        <button v-else class="mt-8 button-lg">Update Competitor</button>
      </form>
    </Layout>
    <not-found v-else />
  </div>
</template>

<script>
import { defineAsyncComponent } from 'vue'
import axios from 'axios'

import Layout from '/@/components/Layout.vue'
import DropdownInput from '/@/components/inputs/DropdownInput.vue'
import TextInput from '/@/components/inputs/TextInput.vue'

const NotFound = defineAsyncComponent(() => import('/@/views/NotFound.vue'))

export default {
  components: {
    Layout,
    DropdownInput,
    TextInput,
    NotFound,
  },

  data: function () {
    return {
      id: '',
      notFound: false,
      create: true,
      leagues: [],
      name: '',
      club: '',
      ageClass: '',
      league: '',
      course: '',
    }
  },

  computed: {
    title: function () {
      if (this.create) return 'Create Competitor'
      else return 'Edit Competitor'
    },

    coursesInLeague: function () {
      const selectedLeague = this.leagues.filter(
        (league) => league.name === this.league
      )
      if (selectedLeague.length > 0) return selectedLeague[0].courses
      else return []
    },
  },

  // On Load
  mounted: function () {
    this.getLeagues()
    if (this.$route.path.includes('edit')) {
      this.create = false
      this.getCompetitorDetails()
    }
  },

  methods: {
    submit: function () {
      if (this.create) this.createCompetitor()
      else this.updateCompetitor()
    },

    getCompetitorDetails: function () {
      return axios
        .get(`/api/competitors/${this.$route.params.id}`)
        .then((response) => {
          if (!response.data) this.notFound = true
          else {
            this.id = this.$route.params.id
            this.name = response.data.name
            this.club = response.data.club
            this.ageClass = response.data.ageClass
            this.league = response.data.league
            this.course = response.data.course
          }
        })
        .catch(() =>
          this.$store.dispatch(
            'createMessage',
            'Problem Getting Competitor Details'
          )
        )
    },

    getLeagues: function () {
      return axios
        .get('/api/leagues')
        .then((response) => {
          this.leagues = response.data
        })
        .catch(() =>
          this.$store.dispatch(
            'createMessage',
            'Problem Fetching List of Leagues'
          )
        )
    },

    validateForm: function () {
      return this.name !== '' && this.league !== '' && this.course !== ''
    },

    createCompetitor: function () {
      if (this.validateForm()) {
        return axios
          .post('/api/competitors', {
            name: this.name,
            club: this.club,
            ageClass: this.ageClass,
            league: this.league,
            course: this.course,
          })
          .then((response) => this.returnToCompetitorsPage(response))
          .catch(() =>
            this.$store.dispatch(
              'createMessage',
              'Error: Problem Creating Competitor - Please Try Again'
            )
          )
      } else
        this.$store.dispatch(
          'createMessage',
          'Please Ensure Name and League Fields are not Blank'
        )
    },

    updateCompetitor: function () {
      if (this.validateForm()) {
        return axios
          .put(`/api/competitors/${this.id}`, {
            id: this.id,
            name: this.name,
            club: this.club,
            ageClass: this.ageClass,
            league: this.league,
            course: this.course,
          })
          .then((response) => this.returnToCompetitorsPage(response))
          .catch((error) =>
            this.$store.dispatch('createMessage', error.response.data.message)
          )
      } else
        this.$store.dispatch(
          'createMessage',
          'Please Ensure Name and League Fields are not Blank'
        )
    },

    returnToCompetitorsPage: function (response) {
      // Go to league page after successful update/ creation
      this.$store.dispatch('createMessage', response.data.message)
      this.$router.push(`/leagues/${this.league}/competitors`)
    },
  },
}
</script>
