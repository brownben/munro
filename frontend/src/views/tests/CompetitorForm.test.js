import { mount } from '@vue/test-utils'
import CompetitorForm from '@/views/CompetitorForm'

test('Is a Vue Instance', () => {
  const wrapper = mount(CompetitorForm, {
    mocks: {
      $route: { path: '' },
    },
  })
  expect(wrapper.isVueInstance()).toBeTruthy()
})

test('Renders Correctly - Create', () => {
  const wrapper = mount(CompetitorForm, {
    mocks: {
      $route: { path: '', params: { id: '' } },
    },
  })
  expect(wrapper.vm.create).toBeTruthy()
  expect(wrapper.element).toMatchSnapshot()
})

test('Renders Correctly - Upload', () => {
  const wrapper = mount(CompetitorForm, {
    mocks: {
      $route: { path: '/competitors/1/edit', params: { id: '' } },
    },
  })
  expect(wrapper.vm.create).toBeFalsy()
  expect(wrapper.element).toMatchSnapshot()
})
