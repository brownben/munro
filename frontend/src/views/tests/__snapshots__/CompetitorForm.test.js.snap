// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Renders Correctly - Create 1`] = `
<div>
  <vue-headful-stub
    title="Munro - Create Competitor"
  />
   
  <div>
    <h1>
      Create Competitor
    </h1>
     
    <!---->
     
    <form>
      <label>
        Name:
      </label>
       
      <input
        type="text"
      />
       
      <label>
        Club:
      </label>
       
      <input
        type="text"
      />
       
      <label>
        Age Class:
      </label>
       
      <input
        type="text"
      />
       
      <label>
        League:
      </label>
       
      <div
        class="dropdown-input"
      >
        <div
          class="visible"
        >
          <label>
            
          </label>
           
          <p>
            
          </p>
           
          <svg
            class=""
            fill="#9E9E9E"
            height="24"
            viewBox="0 0 24 24"
            width="24"
          >
            <path
              d="M7.41 7.84L12 12.42l4.59-4.58L18 9.25l-6 6-6-6z"
            />
             
            <path
              d="M0-.75h24v24H0z"
              fill="none"
            />
          </svg>
        </div>
         
        <div
          class="dropdown"
          name="open"
          style="display: none;"
        />
      </div>
       
      <label>
        Course:
      </label>
       
      <div
        class="dropdown-input"
      >
        <div
          class="visible"
        >
          <label>
            
          </label>
           
          <p>
            
          </p>
           
          <svg
            class=""
            fill="#9E9E9E"
            height="24"
            viewBox="0 0 24 24"
            width="24"
          >
            <path
              d="M7.41 7.84L12 12.42l4.59-4.58L18 9.25l-6 6-6-6z"
            />
             
            <path
              d="M0-.75h24v24H0z"
              fill="none"
            />
          </svg>
        </div>
         
        <div
          class="dropdown"
          name="open"
          style="display: none;"
        />
      </div>
       
      <button>
        Create Competitor
      </button>
       
      <!---->
    </form>
  </div>
   
  <!---->
</div>
`;

exports[`Renders Correctly - Upload 1`] = `
<div>
  <vue-headful-stub
    title="Munro - Edit Competitor"
  />
   
  <div>
    <!---->
     
    <h1>
      Edit Competitor
    </h1>
     
    <form>
      <label>
        Name:
      </label>
       
      <input
        type="text"
      />
       
      <label>
        Club:
      </label>
       
      <input
        type="text"
      />
       
      <label>
        Age Class:
      </label>
       
      <input
        type="text"
      />
       
      <label>
        League:
      </label>
       
      <div
        class="dropdown-input"
      >
        <div
          class="visible"
        >
          <label>
            
          </label>
           
          <p>
            
          </p>
           
          <svg
            class=""
            fill="#9E9E9E"
            height="24"
            viewBox="0 0 24 24"
            width="24"
          >
            <path
              d="M7.41 7.84L12 12.42l4.59-4.58L18 9.25l-6 6-6-6z"
            />
             
            <path
              d="M0-.75h24v24H0z"
              fill="none"
            />
          </svg>
        </div>
         
        <div
          class="dropdown"
          name="open"
          style="display: none;"
        />
      </div>
       
      <label>
        Course:
      </label>
       
      <div
        class="dropdown-input"
      >
        <div
          class="visible"
        >
          <label>
            
          </label>
           
          <p>
            
          </p>
           
          <svg
            class=""
            fill="#9E9E9E"
            height="24"
            viewBox="0 0 24 24"
            width="24"
          >
            <path
              d="M7.41 7.84L12 12.42l4.59-4.58L18 9.25l-6 6-6-6z"
            />
             
            <path
              d="M0-.75h24v24H0z"
              fill="none"
            />
          </svg>
        </div>
         
        <div
          class="dropdown"
          name="open"
          style="display: none;"
        />
      </div>
       
      <!---->
       
      <button>
        Update Competitor
      </button>
    </form>
  </div>
   
  <!---->
</div>
`;
